File: solutions/sorting/alternatives/506-relative-ranks.js
/**

 *
 * This problem demonstrates key concepts in Sorting.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * Assign ranks to athletes based on their scores. The highest score gets "Gold Medal",
 * second gets "Silver Medal", third gets "Bronze Medal", and the rest get their numeric rank.
 * We need to maintain the original order while assigning ranks based on sorted order.
 *
 * APPROACH:





 *
 * WHY THIS WORKS:
 * By pairing each score with its original index, we can sort while maintaining
 * the ability to place results back in the correct positions. Sorting in descending
 * order gives us the natural ranking from highest to lowest.
 *
 * TIME COMPLEXITY: O(n log n) for sorting
 * SPACE COMPLEXITY: O(n) for the result array and sorted pairs
 *
 * EXAMPLE WALKTHROUGH:
 * Input: score = [5,4,3,2,1]
 * Step 1: Create pairs: [[5,0], [4,1], [3,2], [2,3], [1,4]]
 * Step 2: Sort by score (already sorted): [[5,0], [4,1], [3,2], [2,3], [1,4]]
 * Step 3: Assign ranks:
 *   [5,0] -> rank 1 -> "Gold Medal" at index 0
 *   [4,1] -> rank 2 -> "Silver Medal" at index 1
 *   [3,2] -> rank 3 -> "Bronze Medal" at index 2
 *   [2,3] -> rank 4 -> "4" at index 3
 *   [1,4] -> rank 5 -> "5" at index 4
 * Output: ["Gold Medal","Silver Medal","Bronze Medal","4","5"]
 *
 * EDGE CASES:
 * - Single athlete: gets "Gold Medal"
 * - Two athletes: first gets "Gold Medal", second gets "Silver Medal"
 * - Duplicate scores: problem assumes all scores are unique
 */

/**
 * Main solution for Problem 506: Relative Ranks
 *
 * @param {number[]} score - Array of athlete scores
 * @return {string[]} - Array of ranks in original order
 *
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 */
function solve(score) {
    const n = score.length;
    const result = new Array(n);

    // Create pairs of [score, originalIndex]
    const pairs = score.map((s, i) => [s, i]);

    // Sort by score in descending order
    pairs.sort((a, b) => b[0] - a[0]);

    // Assign ranks
    const medals = ["Gold Medal", "Silver Medal", "Bronze Medal"];

    for (let rank = 0; rank < n; rank++) {
        const [scoreValue, originalIndex] = pairs[rank];

        if (rank < 3) {
            result[originalIndex] = medals[rank];
        } else {
            result[originalIndex] = String(rank + 1);
        }
    }

    return result;
}

/**
 * Test cases for Problem 506: Relative Ranks
 */
function testSolution() {
    console.log('Testing 506. Relative Ranks');

    // Test case 1: Example from problem
    const result1 = solve([5, 4, 3, 2, 1]);
    const expected1 = ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"];
    console.assert(JSON.stringify(result1) === JSON.stringify(expected1),
        `Test 1 failed: expected ${JSON.stringify(expected1)}, got ${JSON.stringify(result1)}`);

    // Test case 2: Another example
    const result2 = solve([10, 3, 8, 9, 4]);
    const expected2 = ["Gold Medal", "5", "Bronze Medal", "Silver Medal", "4"];
    console.assert(JSON.stringify(result2) === JSON.stringify(expected2),
        `Test 2 failed: expected ${JSON.stringify(expected2)}, got ${JSON.stringify(result2)}`);

    // Test case 3: Single athlete
    const result3 = solve([100]);
    const expected3 = ["Gold Medal"];
    console.assert(JSON.stringify(result3) === JSON.stringify(expected3),
        `Test 3 failed: expected ${JSON.stringify(expected3)}, got ${JSON.stringify(result3)}`);

    // Test case 4: Two athletes
    const result4 = solve([7, 5]);
    const expected4 = ["Gold Medal", "Silver Medal"];
    console.assert(JSON.stringify(result4) === JSON.stringify(expected4),
        `Test 4 failed: expected ${JSON.stringify(expected4)}, got ${JSON.stringify(result4)}`);

    // Test case 5: Three athletes
    const result5 = solve([1, 2, 3]);
    const expected5 = ["Bronze Medal", "Silver Medal", "Gold Medal"];
    console.assert(JSON.stringify(result5) === JSON.stringify(expected5),
        `Test 5 failed: expected ${JSON.stringify(expected5)}, got ${JSON.stringify(result5)}`);

    console.log('All test cases passed for 506. Relative Ranks!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 506. Relative Ranks ===');
    console.log('Category: Sorting');
    console.log('Difficulty: Easy');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on sorting concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: solutions/sorting/506-relative-ranks.py
"""
# Difficulty: Easy

# 506. Relative Ranks

You are given an integer array score of size n, where score[i] is the score of the ith athlete
in a competition. All the scores are guaranteed to be unique.

The athletes are placed based on their scores, where the 1st place athlete has the highest score,
the 2nd place athlete has the 2nd highest score, and so on. The placement of each athlete determines
their rank:

- The 1st place athlete's rank is "Gold Medal".
- The 2nd place athlete's rank is "Silver Medal".
- The 3rd place athlete's rank is "Bronze Medal".
- For the 4th place to the nth place athlete, their rank is their placement number (i.e., the xth place athlete's rank is "x").

Return an array answer of size n where answer[i] is the rank of the ith athlete.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input description]</dd>
<dt>Output:</dt>
<dd>[output description]</dd>
<dt>Explanation:</dt>
<dd>[explanation]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
We need to map each score to its rank. Sorting gives us the order, but we need to maintain
the original indices. Use sorting with indices or create a score-to-rank mapping.

### APPROACH:
1. **Create index-score pairs**: Track original positions
2. **Sort by score descending**: Highest score first
3. **Assign ranks**: Gold/Silver/Bronze for top 3, numbers for rest
4. **Map back to original positions**: Use original indices

### WHY THIS WORKS:
- Sorting by score gives us the ranking order
- Tracking original indices lets us place ranks correctly
- Dictionary mapping from score to rank is efficient
- Special strings for top 3, numbers for rest

### EXAMPLE WALKTHROUGH:
```
Input: score = [5,4,3,2,1]

Step 1: Create (score, index) pairs
[(5,0), (4,1), (3,2), (2,3), (1,4)]

Step 2: Sort by score descending
[(5,0), (4,1), (3,2), (2,3), (1,4)]

Step 3: Assign ranks
Rank 1 (Gold Medal): score 5, index 0
Rank 2 (Silver Medal): score 4, index 1
Rank 3 (Bronze Medal): score 3, index 2
Rank 4: score 2, index 3
Rank 5: score 1, index 4

Step 4: Map back to original indices
answer[0] = "Gold Medal"
answer[1] = "Silver Medal"
answer[2] = "Bronze Medal"
answer[3] = "4"
answer[4] = "5"

Output: ["Gold Medal","Silver Medal","Bronze Medal","4","5"]
```

### TIME COMPLEXITY:
O(n log n)
For sorting the scores

### SPACE COMPLEXITY:
O(n)
For storing score-rank mappings and result

### EDGE CASES:
- Single athlete (gets Gold Medal)
- Two athletes (Gold and Silver only)
- Three athletes (Gold, Silver, Bronze)
- Large number of athletes

</details>
"""

class Solution:
    def findRelativeRanks(self, score: list[int]) -> list[str]:
        """
        Assign ranks to athletes based on their scores.

        Args:
            score: Array of athlete scores

        Returns:
            Array of rank strings for each athlete

        Time Complexity: O(n log n) for sorting
        Space Complexity: O(n) for result and mappings
        """
        n = len(score)

        # Create (score, original_index) pairs and sort by score descending
        sorted_scores = sorted(enumerate(score), key=lambda x: x[1], reverse=True)

        # Initialize result array
        result = [""] * n

        # Assign ranks
        for rank, (original_idx, _) in enumerate(sorted_scores):
            if rank == 0:
                result[original_idx] = "Gold Medal"
            elif rank == 1:
                result[original_idx] = "Silver Medal"
            elif rank == 2:
                result[original_idx] = "Bronze Medal"
            else:
                result[original_idx] = str(rank + 1)

        return result

    def findRelativeRanksDict(self, score: list[int]) -> list[str]:
        """
        Alternative approach using dictionary mapping.

        Time Complexity: O(n log n)
        Space Complexity: O(n)
        """
        # Create score to rank mapping
        sorted_scores = sorted(score, reverse=True)
        rank_map = {}

        for rank, s in enumerate(sorted_scores):
            if rank == 0:
                rank_map[s] = "Gold Medal"
            elif rank == 1:
                rank_map[s] = "Silver Medal"
            elif rank == 2:
                rank_map[s] = "Bronze Medal"
            else:
                rank_map[s] = str(rank + 1)

        # Map original scores to ranks
        return [rank_map[s] for s in score]

    def findRelativeRanksOneLiner(self, score: list[int]) -> list[str]:
        """
        Concise solution using sorted with key.

        Time Complexity: O(n log n)
        Space Complexity: O(n)
        """
        sorted_scores = sorted(score, reverse=True)
        medals = ["Gold Medal", "Silver Medal", "Bronze Medal"]
        rank_map = {s: medals[i] if i < 3 else str(i + 1) for i, s in enumerate(sorted_scores)}
        return [rank_map[s] for s in score]

def test_solution():
    """Test cases for Problem 506."""
    solution = Solution()

    # Test case 1: Example from problem
    result1 = solution.findRelativeRanks([5, 4, 3, 2, 1])
    expected1 = ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"]
    assert result1 == expected1, f"Expected {expected1}, got {result1}"

    # Test case 2: Another example
    result2 = solution.findRelativeRanks([10, 3, 8, 9, 4])
    expected2 = ["Gold Medal", "5", "Bronze Medal", "Silver Medal", "4"]
    assert result2 == expected2, f"Expected {expected2}, got {result2}"

    # Test case 3: Single athlete
    result3 = solution.findRelativeRanks([100])
    expected3 = ["Gold Medal"]
    assert result3 == expected3, f"Expected {expected3}, got {result3}"

    # Test case 4: Two athletes
    result4 = solution.findRelativeRanks([5, 10])
    expected4 = ["Silver Medal", "Gold Medal"]
    assert result4 == expected4, f"Expected {expected4}, got {result4}"

    # Test case 5: Three athletes
    result5 = solution.findRelativeRanks([1, 2, 3])
    expected5 = ["Bronze Medal", "Silver Medal", "Gold Medal"]
    assert result5 == expected5, f"Expected {expected5}, got {result5}"

    # Test case 6: Four athletes
    result6 = solution.findRelativeRanks([10, 20, 30, 40])
    expected6 = ["4", "Bronze Medal", "Silver Medal", "Gold Medal"]
    assert result6 == expected6, f"Expected {expected6}, got {result6}"

    # Test dictionary approach
    result7 = solution.findRelativeRanksDict([5, 4, 3, 2, 1])
    expected7 = ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"]
    assert result7 == expected7, f"Expected {expected7}, got {result7}"

    # Test one-liner approach
    result8 = solution.findRelativeRanksOneLiner([5, 4, 3, 2, 1])
    expected8 = ["Gold Medal", "Silver Medal", "Bronze Medal", "4", "5"]
    assert result8 == expected8, f"Expected {expected8}, got {result8}"

    # Test case 9: Random order
    result9 = solution.findRelativeRanks([100, 50, 200, 150])
    # Sorted descending: 200, 150, 100, 50
    # Ranks: Gold, Silver, Bronze, 4
    # Original order: [100->Bronze, 50->4, 200->Gold, 150->Silver]
    expected9 = ["Bronze Medal", "4", "Gold Medal", "Silver Medal"]
    assert result9 == expected9, f"Expected {expected9}, got {result9}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print("=== 506. Relative Ranks ===")
    print(f"findRelativeRanks([5,4,3,2,1]) -> {solution.findRelativeRanks([5, 4, 3, 2, 1])}")
    print(f"findRelativeRanks([10,3,8,9,4]) -> {solution.findRelativeRanks([10, 3, 8, 9, 4])}")
    print(f"findRelativeRanks([100]) -> {solution.findRelativeRanks([100])}")
    print(f"findRelativeRanks([1,2,3]) -> {solution.findRelativeRanks([1, 2, 3])}")
